Sources/SwiftSyntaxMacroExpansion/SyntaxProtocol+NonisolatedConformances.swift

```diff
@@ -0,0 +1,147 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements inference of "nonisolated" on the conformances that
+// occur within macro-expanded code. It's meant to provide source compatibility
+//
+
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Given some Swift syntax that may contain type definitions and extensions,
+  /// add "nonisolated" to protocol conformances when there are nonisolated
+  /// members. For example, given:
+  ///
+  ///     extension X: P {
+  ///       nonisolated func f() { }
+  ///     }
+  ///
+  /// this operation will produce:
+  ///
+  ///     extension X: nonisolated P {
+  ///       nonisolated func f() { }
+  ///     }
+  @_spi(Testing) @_spi(Compiler)
+  public func inferNonisolatedConformances() -> Syntax {
+    let rewriter = NonisolatedConformanceRewriter()
+    return rewriter.rewrite(self)
+  }
+}
+
+fileprivate class NonisolatedConformanceRewriter: SyntaxRewriter {
+  override func visitAny(_ node: Syntax) -> Syntax? {
+    // We only care about decl groups (non-protocol nominal types + extensions)
+    // that have nonisolated members and an inheritance clause.
+    guard let declGroup = node.asProtocol(DeclGroupSyntax.self),
+      !declGroup.is(ProtocolDeclSyntax.self),
+      declGroup.containsNonisolatedMembers,
+      let inheritanceClause = declGroup.inheritanceClause
+    else {
+      return nil
+    }
+
+    var skipFirst =
+      declGroup.is(ClassDeclSyntax.self)
+      || (declGroup.is(EnumDeclSyntax.self) && inheritanceClause.inheritedTypes.first?.looksLikeEnumRawType ?? false)
+    let inheritedTypes = inheritanceClause.inheritedTypes.map { inheritedType in
+      // If there's already a 'nonisolated' or some kind of custom attribute
+      if inheritedType.type.hasNonisolatedOrCustomAttribute {
+        return inheritedType
+      }
+
+      if skipFirst {
+        skipFirst = false
+        return inheritedType
+      }
+
+      return inheritedType.with(\.type, "nonisolated \(inheritedType.type)")
+    }
+
+    return Syntax(
+      fromProtocol: declGroup.with(
```

Comment:
```suggestion
      fromProtocol: declGroup.detached.with(
```
so that it doesn't rewrite the whole tree. Same for other `.with(..)`